### DISCUSSION TITLE
Techniques for varying configuration among different machines

### DISCUSSION THREAD

**Discussion by seh** [midchildan/dotfiles#99] [Q&A]
The top-level documentation [advises using a Git topic branch](https://github.com/midchildan/dotfiles#step-3-checkout-the-relevant-branch) (kept presumably on a given machine) in order to adjust the common configuration for that machine. There's also [mention there](https://github.com/midchildan/dotfiles#step-2-setup-your-repository) of editing the _config.toml_ file.

While it's easy to understand that there could be these separate copies of the code on several machines, I'm wondering if you ever go back and commit more machines to [the _nix/home/machines/default.nix_ file](https://github.com/midchildan/dotfiles/blob/152b40c3a412b18ba6057c3ecfb984748962282b/nix/home/machines/default.nix#L7-L12). It may not be of any use to publish this machine-specific configuration to the common source code, as other machines shouldn't need to use it, but I can see some benefit of being able to get started with a new machine by cloning the repository and then saying, "This new machine _B_ is just like this other machine _A_ which is already defined here, except for a couple of details that need adjustment."

Also, say that I wanted to vary only a detail like the email address that I'll use, and maybe the GPG signing key ID, but the rest of the _config.toml_ file should stay the same. Can I override the _config.toml_ file's values elsewhere, such as in a _flake.nix_ file? Could I assign to an attribute like `dotfiles.config.user.email` in a module passed to `lib.mkDarwin` like you show [in the template _flake.nix_ file](https://github.com/midchildan/dotfiles/blob/1c190d0ac1d87c159b8b7d777f02261ae58a3fc5/nix/templates/darwin/flake.nix#L19-L29)?

Again, I understand how maintaining a fork of the _config.toml_ would work, but I'm interested in other ways to override parts of that configuration for each machine that would all be committed together in the same source code repository (and not just on parallel branches in that repository).

**Comment by midchildan** [comment]
> While it's easy to understand that there could be these separate copies of the code on several machines, I'm wondering if you ever go back and commit more machines to [the nix/home/machines/default.nix file](https://github.com/midchildan/dotfiles/blob/152b40c3a412b18ba6057c3ecfb984748962282b/nix/home/machines/default.nix#L7-L12).

I use git branches for non-Nix configuration. Nix configuration is kept the same across all branches because the module system is really good at handling differences between machines in a sane way. I keep configuration for all of my machines in a single branch too, but they're in a separate private repository.

> "This new machine B is just like this other machine A which is already defined here, except for a couple of details that need adjustment."

When I have a situation like this, I tend to make a new module under [the profiles directory](https://github.com/midchildan/dotfiles/tree/main/nix/home/profiles) and expose the differences as module options.

> Can I override the config.toml file's values elsewhere, such as in a flake.nix file?

This isn't currently possible. However, since `config.toml` just provides the defaults for some NixOS / Home Manager / nix-darwin options so you can just override them individually.

I've been considering introducing the modules system to flakes using [flake-parts](https://flake.parts). This should make it possible to implement what you described, but I haven't decided on that yet.

**Reply by seh** [reply]
With _flake-parts_ in play, I see that what had been in the _nix/config.toml_ file moved into default values in the _nix/config.nix_ file's option definitions. Have you had the occasion to try overriding those default values for different machines or environments?

It's obvious to me that I can edit the default values in place, but it looks like it's supposed to be possible to specify alternate values from _somewhere_. Would those alternate values sit in the top-level _flake.nix_ file?

**Reply by midchildan** [reply]
> Have you had the occasion to try overriding those default values for different machines or environments?

Yes, I used the `home.modules`, `nixos.modules`, and `darwin.modules` option to add more modules for Home Manager, NixOS, and nix-darwin in a separate private repo. But since the flake-parts options are meant as flake-wide defaults, I just stick to Home Manager / NixOS / nix-darwin options when configuring individual machines.

> It's obvious to me that I can edit the default values in place, but it looks like it's supposed to be possible to specify alternate values from somewhere. Would those alternate values sit in the top-level flake.nix file?

It's indeed possible to override the default values in `flake.nix` or any imported flake-parts module. However, the intent is to offer customization options to external repositories that import this dotfiles as a flake-parts module.

**Reply by seh** [reply]
> But since the flake-parts options are meant as flake-wide defaults, I just stick to Home Manager / NixOS / nix-darwin options when configuring individual machines.

To take the "user.name" option defined in file _nix/config.nix_ as an example, that one is taken into account in several places: _nix-darwin_ configuration, _Home Manager_ configuration, as well as configurations for programs like _git_ and _gpg_. I'd prefer to be able to override it _once_—to avoid repeating myself—for a given _nix-darwin_ machine.

It sounds like you don't do that within this same repository, though, and that you defined this option for use by other flakes importing this repository as a _flake-parts_ module. Is that correct?

I'm still trying to figure out the right place to override these default values.

**Reply by seh** [reply]
I've been trying to adjust the user name and email address values by passing a module into [the `importDarwin` call in file _nix/darwin/machines/default.nix_](https://github.com/midchildan/dotfiles/blob/7f3eadda1c3ef66fe81919cad0ca85b1680a5956/nix/darwin/machines/default.nix#L10), albeit in my own repository using a different name for a different machine definition.

Still, just groping around, I've tried something like this:
```nix
{ inputs, ... }:

let
  inherit (inputs.self.lib) importDarwin pkgsFor;
in
{
  flake.darwinConfigurations = {
    local =
      importDarwin ./work.nix
        {
          pkgs = pkgsFor "aarch64-darwin";
          modules = [
            (
              { ... }:
              {
                # None of these attempts to change the "fullName" attribute have made in through
                # into the flake's outputs.
                config.dotfiles._flakeOptions.user.fullName = "AAA";
                dotfiles._flakeOptions.user.fullName = "BBB";
                dotfiles.user.fullName = "CCC";
              }
            )
          ];
        };
  };
}
```
That winds up defining a machine named "local", and I can use it, but the "fullName" attribute value remains the same as the default value.

I'm sure that these attempts so far do _not_ work, but this seems like the place at which I'd like to be able to override the default values, as I'm defining some machine on which I know that my user name and email address are different from my normal values (such as for my primary computer that I use for my job).

**Reply by seh** [reply]
Oh, I realized that that attempt to set the "modules" attribute is ignored by `importDarwin`, which [sets the attribute unconditionally](https://github.com/midchildan/dotfiles/blob/7f3eadda1c3ef66fe81919cad0ca85b1680a5956/nix/lib/flake-module.nix#L98) to `(import configPath)`. Using [`mkDarwin`](https://github.com/midchildan/dotfiles/blob/7f3eadda1c3ef66fe81919cad0ca85b1680a5956/nix/lib/flake-module.nix#L59-L93) directly allows adding more modules.

Now I'm struggling with how to update a value in the "dotfiles._flakeOptions.user" attribute set from within such a module.

**Reply by midchildan** [reply]
> It sounds like you don't do that within this same repository, though, and that you defined this option for use by other flakes importing this repository as a flake-parts module. Is that correct?

Yes, that's correct.

> I realized that that attempt to set the "modules" attribute is ignored by importDarwin

With `importDarwin`, modules can be imported through the file specified by `configPath`.

> Now I'm struggling with how to update a value in the "dotfiles._flakeOptions.user" attribute set from within such a module.

This isn't possible because I defined the option in a way such that [it can't be set in multiple places](https://github.com/midchildan/dotfiles/blob/7f3eadda1c3ef66fe81919cad0ca85b1680a5956/nix/darwin/default.nix#L28). So currently, the options are:

1. Set options that use `dotfiles.flakeOptions.user.fullName` (e.g., `dotfiles.git.user.name`)
2. Create a new git branch and update `nix/config.nix`.
